libraries/cosmic_unicorn/cosmic_unicorn.cpp

```diff
@@ -0,0 +1,578 @@
+      dma_safe_abort(audio_dma_channel);
+
+      play_mode = NOT_PLAYING;
+    }
+  }
+
+  AudioChannel& CosmicUnicorn::synth_channel(uint channel) {
+    assert(channel < PicoSynth::CHANNEL_COUNT);
+    return synth.channels[channel];
+  }
+
+  void CosmicUnicorn::set_pixel(int x, int y, uint8_t r, uint8_t g, uint8_t b) {
+    if(x < 0 || x >= WIDTH || y < 0 || y >= HEIGHT) return;
+
+    x = (WIDTH - 1) - x;
+    y = (HEIGHT - 1) - y;
+
+    // map coordinates into display space
+    if(y < 16) {
+      // move to top half of display (which is actually the right half of the framebuffer)
+      x += 32;
+    }else{
+      // remap y coordinate
+      y -= 16;      
+    }
+
+    r = (r * this->brightness) >> 8;
+    g = (g * this->brightness) >> 8;
+    b = (b * this->brightness) >> 8;
+
+    uint16_t gamma_r = r_gamma_lut[r];
+    uint16_t gamma_g = g_gamma_lut[g];
+    uint16_t gamma_b = b_gamma_lut[b];
+
+    // for each row:
+    //   for each bcd frame:
+    //            0: 00111111                           // row pixel count (minus one)
+    //      1  - 64: xxxxxbgr, xxxxxbgr, xxxxxbgr, ...  // pixel data
+    //      65 - 67: xxxxxxxx, xxxxxxxx, xxxxxxxx       // dummy bytes to dword align
+    //           68: xxxxrrrr                           // row select bits
+    //      69 - 71: tttttttt, tttttttt, tttttttt       // bcd tick count (0-65536)
+    //
+    //  .. and back to the start
+
+    // set the appropriate bits in the separate bcd frames
+    for(uint8_t frame = 0; frame < BCD_FRAME_COUNT; frame++) {
+      uint8_t *p = &bitstream[y * ROW_BYTES + (BCD_FRAME_BYTES * frame) + 1 + x];
+
+      uint8_t red_bit = gamma_r & 0b1;
+      uint8_t green_bit = gamma_g & 0b1;
+      uint8_t blue_bit = gamma_b & 0b1;
+
+      *p = (blue_bit << 0) | (green_bit << 1) | (red_bit << 2);
+
+      gamma_r >>= 1;
+      gamma_g >>= 1;
+      gamma_b >>= 1;
+    }
+  }
+
+  void CosmicUnicorn::set_brightness(float value) {
+    value = value < 0.0f ? 0.0f : value;
+    value = value > 1.0f ? 1.0f : value;
+    this->brightness = floor(value * 256.0f);
+  }
+
+  float CosmicUnicorn::get_brightness() {
+    return this->brightness / 255.0f;
+  }
+
+  void CosmicUnicorn::adjust_brightness(float delta) {
+    this->set_brightness(this->get_brightness() + delta);
+  }
+
+  void CosmicUnicorn::set_volume(float value) {
+    value = value < 0.0f ? 0.0f : value;
+    value = value > 1.0f ? 1.0f : value;
+    this->volume = floor(value * 255.0f);
+    this->synth.volume = this->volume * 255.0f;
+  }
+
+  float CosmicUnicorn::get_volume() {
+    return this->volume / 255.0f;
+  }
+
+  void CosmicUnicorn::adjust_volume(float delta) {
+    this->set_volume(this->get_volume() + delta);
+  }
+
+  void CosmicUnicorn::update(PicoGraphics *graphics) {
+    if(unicorn == this) {
+      if(graphics->pen_type == PicoGraphics::PEN_RGB888) {
+        uint32_t *p = (uint32_t *)graphics->frame_buffer;
+        for(size_t j = 0; j < 32 * 32; j++) {

```

Comment:
We should just loop through y and x values here, since we already know their ranges and calculating them from `j` is unnecessarily costly.